File: Server/lobby.py
import uuid

class Member:
    def __init__(self,sid) -> None:
        self.sid = sid
        self.ready = False

class Lobby:
    def __init__(self) -> None:
        self.connected = {}
        self.team_one = []
        self.team_two = []
        self.team_one_ready = 0
        self.team_two_ready = 0

    def is_full(self):
        return len(self.connected) >= 4

    def add_player(self,sid):
        self.connected[sid] = Member(sid)

    def remove_player(self,sid):
        self.connected.pop(sid)
        self.team_one = [m for m in self.team_one if m.sid != sid]
        self.team_two = [m for m in self.team_two if m.sid != sid]
        print(self.team_one,self.team_two)

    def update_ready(self,ready,team,sid):

        if team == 1:
            team_one_count = 0

            for m in self.team_one:
                if m.sid == sid:
                    m.ready = ready
                if m.ready:
                    team_one_count += 1
            self.team_one_ready = team_one_count
        else:
            team_two_count = 0

            for m in self.team_two:
                if m.sid == sid:
                    m.ready = ready

                if m.ready:
                    team_two_count +=1 
            self.team_two_ready = team_two_count
        
        return self.team_one_ready + self.team_two_ready
    
    def is_ready(self):
        return (self.team_two_ready + self.team_one_ready) == 4
    
    def join_team(self,team, sid):
        if team == 1:
            self.team_one.append(self.connected[sid])
        else:
            self.team_two.append(self.connected[sid])
        print("Count",len(self.team_one),len(self.team_two))

    def leave_team(self,team,sid):
        if team == 1:
            self.team_one = [m for m in self.team_one if m.sid != sid]
        else:
            self.team_two = [m for m in self.team_two if m.sid != sid]
        print("Count",len(self.team_one),len(self.team_two))

File: Client/src/Components/Lobby.js
import { Button } from "@mui/material";
import { display } from "@mui/system";
import { useState, useEffect } from "react";
import { v4 as uuidv4 } from 'uuid';


function Lobby(props){
    const [teamOne,setTeamOne] = useState([])
    const [teamTwo,setTeamTwo] = useState([])
    const [ready,setReady] = useState(false)
    const [currentTeam,setCurrentTeam] = useState("")
    const [totalReady,setTotalReady] = useState(0)
    const addPlayer = (data)=>{
        console.log("added player")
        if ( data["team"] === 1){
            
            setTeamOne(prevTeamOne => {
                console.log(prevTeamOne)
                const n =[
                    ...prevTeamOne,
                    data["username"]
                  ]
                console.log(n)
                return n  })
        }
        else{
            setTeamTwo(prevTeamTwo => {
                return [
                  ...prevTeamTwo,
                  data["username"]
                ]})
            console.log(teamTwo)
        } 
    }

    const removePlayer = (data) =>{
        console.log("Remove player", data)
        
        if (data["team"] === 1){
            setTeamOne( prev => { const n = prev.filter(p => p !== data["username"]) 
                console.log(n,data["username"])
                return n    
            })
        }
        else{
            setTeamTwo( prev => { const n = prev.filter(p => p !== data["username"]) 
                console.log(n, data["username"])
                return n    
            })
        }
    }
    const selectTeam = (team) =>{
        console.log("clicked",team)
        
        if( currentTeam === ""){
            console.log("here")
            props.lobbyApi.selectTeam(team)
            setCurrentTeam(team)
        }
        else if (currentTeam !== team){
            console.log(props.username)
            props.lobbyApi.leaveTeam(currentTeam,props.username);
            props.lobbyApi.selectTeam(team)
            setCurrentTeam(team)

        }
    }

    const setInitialTeams = (teamOne,teamTwo) =>{
        setTeamOne(teamOne)
        setTeamTwo(teamTwo)
    }
    useEffect( () => {
        props.lobbyApi.getTeams(setInitialTeams)
        props.lobbyApi.onTeamSelection(addPlayer)
        props.lobbyApi.onLeft(removePlayer)
        props.lobbyApi.onTotalReady(setTotalReady)
        props.lobbyApi.onStart(() => props.navigate("Table"))
    },[])
    
    const buttons = () =>{
        if(!ready){
            return (
            <div>
                <Button onClick={()=>selectTeam(1)} variant="outlined" >Join Team One</Button>
                <Button onClick={() => selectTeam(2)} variant="outlined">Join Team Two</Button>
            </div>)
        }
        else{
            return <div></div>
        }
    }
    const readyText = () =>{
        if(ready){
            return "Un-ready"
        }
        else{
            return "Ready"
        }
    }

    return (
        <div>
            <h3>Lobby</h3>
            { buttons()}
            <Button onClick={() => {
                const r = (!ready && currentTeam !== "" && ( currentTeam == 1 ? (teamOne.length <=2) : (teamTwo.length <=2)))
                if( r !== ready){
                    props.lobbyApi.updateReady(r)
                    setReady(r)
                }
                }}>{readyText()}</Button>
            <p>{totalReady}</p>
            <div style={{display:'flex', justifyContent:'space-between' }}>
                <div style={{padding:'15px'}}>
                    <p>Team One</p>
                    {teamOne.map((item, index) => (
                        <li key={index}>{item}</li>
                        ))}
                </div>
                
                <div style={{padding:'15px'}}>
                    <p>Team Two</p>
                    {teamTwo.map((item, index) => (
                        <li key={index}>{item}</li>
                        ))}
                </div>
            </div>
        </div>
    )
}
export default Lobby;